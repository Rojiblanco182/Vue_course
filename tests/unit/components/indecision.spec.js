import { shallowMount } from "@vue/test-utils"
import Indecision from '@/components/Indecision'

let wrapper
global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve({
        answer: 'yes',
        forced: false,
        image: 'mockUrl'
    })
}))

beforeEach(() => {
    wrapper = shallowMount(Indecision)

    jest.clearAllMocks()
})

describe('Indecision component', () => {
    test('should match the snapshot created', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    
    test('should only trigger a console.log when we are still typing', async () => {
        const input = wrapper.find('input')
        const spyOnConsole = jest.spyOn(console, 'log')
        const spyOnGetAnswer = jest.spyOn(wrapper.vm, 'getAnswer')
        await input.setValue('Mock value')

        expect(spyOnConsole).toHaveBeenCalledTimes(1)
        expect(spyOnGetAnswer).not.toHaveBeenCalled()
    })
    
    test('should call getAnswer when we type the interrogation mark', async () => {
        const input = wrapper.find('input')
        const spyOnGetAnswer = jest.spyOn(wrapper.vm, 'getAnswer')

        await input.setValue('Will this test pass?')

        expect(spyOnGetAnswer).toHaveBeenCalledTimes(1)
    })
    
    test('getAnswer should run properly', async () => {
        await wrapper.vm.getAnswer()

        const image = wrapper.find('img')

        expect(image.exists()).toBeTruthy()
        expect(wrapper.vm.imgSource).toBe('mockUrl')
        expect(wrapper.vm.answer).toBe('SÃ­!')
    })
    
    test('should return an error message if there was an error calling the API', async () => {
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))
        await wrapper.vm.getAnswer()

        const image = wrapper.find('img')

        expect(image.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo obtener una respuesta')
    })
    
})