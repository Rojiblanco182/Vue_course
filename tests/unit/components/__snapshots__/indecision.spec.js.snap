// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component should match the snapshot created 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar la pregunta con ?</p>
  <!--v-if-->
</div>
`;
